src/error.rs

```diff
@@ -28,51 +30,139 @@ macro_rules! write_err {
 
 /// Hex decoding error.
 #[derive(Debug, Clone, PartialEq, Eq)]
-pub enum HexToBytesError {
-    /// Non-hexadecimal character.
-    InvalidChar(InvalidCharError),
+pub enum FromHexError<E> {
```

Comment:
This is an alternate approach to how we currently have it, using the `From` trait bound on `Self::Error`.

`examples::custom` has an example of how I thought this would be used:
```rust
impl FromHex for ALittleBitHexy {
    type Error = CustomError;

    fn from_byte_iter<I>(iter: I) -> Result<Self, Self::Error>
    where
        I: Iterator<Item = Result<u8, InvalidCharError>> + ExactSizeIterator + DoubleEndedIterator,
    {
        // Errors if the iterator is the wrong length.
        let data = <[u8; 32] as FromHex>::from_byte_iter(iter).map_err(CustomError::Hex)?;

        // An example of some application specific error.
        if data == [0; 32] {
            return Err(CustomError::AllZeros);
        }

        // This is a contrived example (using x==0).
        Ok(ALittleBitHexy { data, x: 0 })
    }
}
```

> Shouldn't it be the other way around? Have InvalidChar variant and generic InvalidLength variant where the InvalidLength would be either OddLengthError or InvalidLengthError.

I didn't fully process this bit. I can have a play with this idea if the custom example above doesn't change you view.